Feature 929 genomics correlation

asizemore opened the pull request:
Resolves #929 

This PR does two things:
1. Updates the api for all correlation apps to respect new api ( see https://github.com/VEuPathDB/service-eda-compute/pull/90)
2. Adds a `correlation.tsx` plugin to be used for genomics (- vb) sites. 

Open todos:

- [x] Ensure user can't select two of the same input
- [x] Select metadata as Data 2
- [x] Remove old hacks for the original two correlation plugins that were just there to give genomics some flexibility.
- [x] Clean up and think again about the new component

Some things I thought about while writing:
- How to append this new "metadata" fake collection to the items in the `VariableCollectionSingleSelect` dropdown. For a while i made a new component called "MixedVariableSingleSelect" that would ideally handle a mixture of both collection and regular variables. I don't know where else we'd use such a component though... anyways I'm not convinced what i've done is the best, so please take a look and let me know if you have other ideas.
- I thought about extending `CorrelationConfig` to be `CorrelationAssayAssayConfig` and `CorrelationAssayMetadataConfig` with the latter two having the `data2` prop set already to definitely be a collection or metadata. I started into the weeds on this but got confused, so if you, reviewer, would like to chat about it i'd like to learn!


**Note** This PR requires a local backend setup with the `correlations-refactor` branch of the compute and data service.

asizemore commented:
Update:
Encountering more challenges than expected. If the user chooses "metadata" from the dropdown, the request should contain simply nothing in the `collectionVariable2` slot. However, trying to select an item from the SelectList with an undefined value got me into trouble because the component also needs to know when nothing is selected so it can say "Select a blah". Then i thought i'll put in a fake value and then intercept that fake value in  `handleSelect` so that it doesn't actually get added to the compute config. That would work for the viz page but then on the thumbnail page it wouldn't know to show Data 2: Metadata or Data 2: _Not selected_ , because for both the computeConfig would have nothing for `collectionVariable2`.

asizemore commented:
Actually this is more complicated than yesterday me thought. If Data 2 is basically optional, then the compute config is perfectly valid with no Data 2 value selected.

asizemore commented:
just found a bug hang on... taking it out of ready for reveiw until it's resolved...

asizemore commented:
Okay it's ready. I had more double `changeConfigHandler` issues. It's possible i should just update that function to take an array of prop names....

jernestmyers commented:
So I've done a cursory review of the code and things look good to me. When the backend is ready, perhaps we could sync up before a deeper review and testing?

asizemore commented:
@jernestmyers Thanks for looking through the code! Yes that's a good idea. I'll keep watch on those backend branches and let you both know when we're good to go for a sync

asizemore commented:
@jernestmyers and @dmfalke just an update:
1. Dave's PR for node clicking abilities has been merged, so i'm going to work on incorporating those new features into this branch
2. The backend should be merged today

So i'll reach back out to both of you when 1 and 2 are done and we can do a real review!

dmfalke commented on packages/libs/eda/src/lib/core/components/computations/plugins/correlationAssayMetadata.tsx:
```diff
@@ -42,31 +43,16 @@ const cx = makeClassNameHelper('AppStepConfigurationContainer');
     CorrelationAssayMetadataConfigDescriptionComponent,
   createDefaultConfiguration: () => ({}),
-  isConfigurationComplete: CompleteCorrelationAssayMetadataConfig.is,
+  isConfigurationComplete: CompleteCorrelationConfig.is,
```
Will `data2` be `undefined` for this app? If so, not sure this io-ts type will work, since it will enforce that all properties have a value.

dmfalke reviewed (commented):
I've skimmed the code, and nothing stands out to me, except for the one inline comment I left.

I will work on testing the code shortly.

dmfalke commented:
I'm having trouble running services locally. I suspect it's related to the consolidation. I will try to get some help later today, or tomorrow.

dmfalke commented:
> I'm having trouble running services locally. I suspect it's related to the consolidation. I will try to get some help later today, or tomorrow.

Figured it out. I was using the wrong image tag name

dmfalke reviewed (commented):
In the consolidated app, if I choose metadata for data2, the config is not recognized as complete.

![image](https://github.com/VEuPathDB/web-monorepo/assets/365139/76d143ce-78dd-4d5e-992e-e0cfb45c21cc)

dmfalke commented:
I'm getting a 500 for the compute job in mbio:
![image](https://github.com/VEuPathDB/web-monorepo/assets/365139/19daa5fe-c6bc-40c0-8123-ebd076cc010c)

asizemore commented:
Thanks for your testing, @dmfalke ! Good catch on the metadata - that value changed one more time in the api since i last was here in the frontend. I'll fix that up.

And all the computes failing is expected for right now until the backend is all merged

asizemore commented on packages/libs/eda/src/lib/core/components/computations/plugins/correlationAssayMetadata.tsx:
```diff
@@ -145,18 +131,24 @@ export function CorrelationAssayMetadataConfiguration(
+  configuration.data2 = {
+    dataType: 'metadata',
+  };
+
```
oh it absolutely should! Fixed 👍

dmfalke replied:
Great!

In the same vein, can this also be moved there?

https://github.com/VEuPathDB/web-monorepo/blob/817d73261af25314deb731c1bf3f21c09bd3fd3a/packages/libs/eda/src/lib/core/components/computations/plugins/correlationAssayMetadata.tsx#L160-L173

asizemore commented:
> In the consolidated app, if I choose metadata for data2, the config is not recognized as complete.

This turned out to be because of how I had defined `isCompleteConfiguration`. I had added a check to ensure that the collection inputs were unique, but had not added logic to allow for the metadata case to pass through. Good catch! The updated logic is in and fixes the issue. It probably could have been condensed a little but it was getting hard to read so i split it out over a few lines for clarity.

asizemore commented:
Allrighty. All updated! I've addressed Dave's feedback and also updated this branch with `main`. Feel free to have another look!

You can use my dev site as a backend if you need while the qa site is down. Looks like the backend isn't merged yet anyhow.

dmfalke commented:
> And all the computes failing is expected for right now until the backend is all merged

I was using a local backend with the branch. Maybe I forgot to point my local dev site to it.

dmfalke reviewed (commented):
I think this all looks great. One final question I have is related to default configurations, re `prefilterThresholds`. Seems like these could always be set in `createDefaultConfiguration`.

asizemore commented:
@dmfalke I just moved the `prefilterThresholds`. Good idea!

I have to run so i haven't tested yet, but i'll do that first thing monday morning. Looks like the backend merges are coming along anyways 🎉

dmfalke commented:
Did a quick test, using my dev site for a backend. The front end code appears to be working as expected. However, I'm getting an empty response for the bpnet data:

**Request**

![image](https://github.com/VEuPathDB/web-monorepo/assets/365139/c4b2c1ab-232c-4031-a333-6f36fbd1ce6c)

**Response**

![image](https://github.com/VEuPathDB/web-monorepo/assets/365139/ddbd3fd4-3c03-406a-9af3-6b1c4dd9dbe6)

asizemore commented:
@dmfalke that's so odd! I'll look into it this morning.

asizemore commented:
I'm seeing all the computes failing similarly (see https://github.com/VEuPathDB/service-eda/issues/7). I'm going to start a local backend pre-consolidation and see if that works

asizemore commented:
update. Confirmed the lack of data being returned is a backend issue. There's an api mismatch between compute and data service. I'll be working on it tomorrow.

asizemore commented:
Update. All working again! Please use the `fix-viz-plugin-correlation-api` branch of the consolidated backend.

asizemore commented:
<img width="1225" alt="Screen Shot 2024-04-17 at 7 56 38 AM" src="https://github.com/VEuPathDB/web-monorepo/assets/11710234/b32e20c0-3643-4fb7-b48f-28bc23433cc7">

dmfalke reviewed (approved):
I think the client code is ready to go! I'm still having trouble getting a plot in mbio, but that could be for any variety of reasons, and does not seem to be specific to the client code changes.

Unless you feel that there is anything missing, I would go ahead and merge this and get the changes to qa so they can be further tested.

Great work!!
